File: homework_four_dot_one/src/main/java/com/abra/homework_four_dot_one/ContactItemAdapter.java
package com.abra.homework_four_dot_one;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class ContactItemAdapter extends RecyclerView.Adapter<ContactItemAdapter.ContactItemViewHolder> implements Filterable {
    private final ArrayList<ContactItem> contactItemList;
    private final ArrayList<ContactItem> contactItemListFull;
    private OnItemClickListener onItemClickListener;

    public ArrayList<ContactItem> getContactItemList() {
        return contactItemList;
    }

    public ContactItemAdapter(ArrayList<ContactItem> contactItemList) {
        this.contactItemList = contactItemList;
        contactItemListFull = new ArrayList<>(contactItemList);
    }

    @NonNull
    @Override
    public ContactItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_contact_info, parent, false);
        return new ContactItemViewHolder(view, onItemClickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull ContactItemViewHolder holder, int position) {
        holder.bind(contactItemList.get(position));
    }

    @Override
    public int getItemCount() {
        return contactItemList != null ? contactItemList.size() : 0;
    }

    public void add(ContactItem contactItem) {
        contactItemList.add(contactItem);
        contactItemListFull.add(contactItem);
        notifyItemChanged(contactItemList.indexOf(contactItem));
    }

    public void edit(int position, ContactItem contactItem) {
        contactItemList.set(position, contactItem);
        contactItemListFull.set(position, contactItem);
        notifyItemChanged(position);
    }

    public void remove(int position) {
        contactItemList.remove(position);
        contactItemListFull.remove(position);
        notifyItemRemoved(position);
    }

    @Override
    public Filter getFilter() {
        return filter;
    }

    private final Filter filter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            List<ContactItem> filteredList = new ArrayList<>();
            if (charSequence == null || charSequence.length() == 0) {
                filteredList.addAll(contactItemListFull);
            } else {
                String filterPattern = charSequence.toString().toLowerCase().trim();
                for (ContactItem item : contactItemListFull) {
                    if (item.getName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(item);
                    }
                }
            }
            FilterResults results = new FilterResults();
            results.values = filteredList;
            return results;
        }

        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            contactItemList.clear();
            contactItemList.addAll((ArrayList) filterResults.values);
            notifyDataSetChanged();
        }
    };

    static class ContactItemViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageView;
        private final TextView name;
        private final TextView info;
        private final OnItemClickListener onItemClickListener;

        public ContactItemViewHolder(@NonNull View itemView, OnItemClickListener onItemClickListener) {
            super(itemView);
            imageView = itemView.findViewById(R.id.imageView);
            name = itemView.findViewById(R.id.tvName);
            info = itemView.findViewById(R.id.tvInfo);
            this.onItemClickListener = onItemClickListener;
        }

        void bind(ContactItem contactItem) {
            imageView.setImageResource(contactItem.getIconId());
            switch (contactItem.getTypeInfo()) {
                case "phone": {
                    imageView.setBackgroundResource(R.drawable.shape2);
                    break;
                }
                case "email": {
                    imageView.setBackgroundResource(R.drawable.shape1);
                    break;
                }
            }
            if (onItemClickListener != null)
                itemView.setOnClickListener(view -> {
                    onItemClickListener.onItemClick(contactItem, getAdapterPosition());
                });
            name.setText(contactItem.getName());
            info.setText(contactItem.getInfo());
        }
    }

    interface OnItemClickListener {
        void onItemClick(ContactItem contactItem, int position);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }
}

File: homework_8_1/src/main/java/com/abra/homework_8_1/ContactItemAdapter.kt
package com.abra.homework_8_1

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import java.util.Locale
import kotlin.collections.ArrayList

class ContactItemAdapter : RecyclerView.Adapter<ContactItemAdapter.ContactItemViewHolder>(), Filterable {
    private var contactItemList = arrayListOf<ContactItem>()
    private var contactItemListFull = arrayListOf<ContactItem>()
    private var contactItemListFullBeforeUsingSearchView = arrayListOf<ContactItem>()
    lateinit var itemClickListener: (ContactItem, Int) -> Unit

    /*
    * Эти функции мне понадобились для получения актуального листа,
    * вторая нужна, для восстановления листа после использования SearchView,
    * так как с ним возникали багги.
    *
    * P.S Честно скажу, что в домашке 4.1 я так и
    * не исправил эти баги:), но тут все же решился разобраться.
    * */
    fun getList() = contactItemList
    fun getListBeforeUsingSearchView() = contactItemList

    class ContactItemViewHolder(private val itemView: View, private val itemClickListener: (ContactItem, Int) -> Unit) : RecyclerView.ViewHolder(itemView) {
        private var imageView = itemView.findViewById<ImageView>(R.id.imageView)
        private var name = itemView.findViewById<TextView>(R.id.tvName)
        private var info = itemView.findViewById<TextView>(R.id.tvInfo)
        fun bind(item: ContactItem) {
            imageView.setImageResource(item.iconId)
            when (item.typeInfo) {
                "phone" -> imageView.setBackgroundResource(R.drawable.shape2)

                "email" -> imageView.setBackgroundResource(R.drawable.shape1)
            }
            itemView.setOnClickListener {
                itemClickListener.invoke(item, adapterPosition)
            }
            name.text = item.name
            info.text = item.info
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_contact_info, parent, false)
        return ContactItemViewHolder(view, itemClickListener)
    }

    override fun onBindViewHolder(holder: ContactItemViewHolder, position: Int) {
        holder.bind(contactItemList[position])
    }

    override fun getItemCount() = contactItemList.size

    private val filter: Filter = object : Filter() {
        override fun performFiltering(p0: CharSequence?): FilterResults {
            val filteredList = arrayListOf<ContactItem>()
            if (p0 == null || p0.isEmpty()) {
                filteredList.addAll(contactItemListFull)
            } else {
                val filterPattern = p0.toString().toLowerCase(Locale.ROOT).trim()
                contactItemListFull.forEach {
                    if (it.name.toLowerCase(Locale.ROOT).contains(filterPattern)) {
                        filteredList.add(it)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        override fun publishResults(p0: CharSequence?, p1: FilterResults?) {
            contactItemList.clear()
            contactItemList.addAll(p1?.values as ArrayList<ContactItem>);
            notifyDataSetChanged()
        }
    }

    override fun getFilter() = filter

    /*
    * Решил поступить таким способом, что мне кажется выглядит красивее,
    * нежели я бы добавил методы по удалению, редактированию и добавлению.
    * К тому же пришлось бы туда-сюда передавать лишнюю информацию.
    * */
    fun updateList(list: ArrayList<ContactItem>) {
        contactItemList = list
        contactItemListFull = ArrayList(contactItemList)
        contactItemListFullBeforeUsingSearchView = ArrayList(contactItemList)
        notifyDataSetChanged()
    }
}